### PR TITLE
i-bem__dom: update _elemCache on findElem calls

### LINKED ISSUE #347: i-bem. this.elem cache
this.elem cache should be dropped after findElem call
also is it possible to drop elem cache after jQuery dom manipulation functions calls? like empty or append, remove?

### PR DESCRIPTION
**Pull request by qfox**
Partially fixes #347
Port of bem/bem-core#842

### CODE REVIEW

**Comment by qfox** [comment]
Here we don't have specs but this code tested in usual way in bem-core.

**Comment by deeonis** [comment]
we'll look closely next week: BEM-2298

File: blocks-common/i-bem/__dom/i-bem__dom.js [hunk 1]
```diff
@@ -805,11 +805,23 @@ var DOM = BEM.DOM = BEM.decl('i-bem__dom', {
...
+            res = findDomElem(ctx, selector);
+
+        // caching results if possible
+        (ctx.length === 1 && this.domElem.length === 1 && ctx[0] === this.domElem[0]) &&
+            keys.forEach(function(key, i) {
+                (this._elemCache[key] = isSingleName? res : res.filter('.' + key))
+                    .__bemElemName = names[i];
+            }, this);
+
         return findDomElem(ctx, selector);
```
Comment (Yeti-or) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: @zxqfox  Let's return `res` value instead of another call of findDomElem ?

Comment (qfox) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: oh, right!

Comment (qfox) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: @Yeti-or fixed

File: blocks-common/i-bem/__dom/i-bem__dom.js [hunk 2]
```diff
@@ -805,11 +805,23 @@ var DOM = BEM.DOM = BEM.decl('i-bem__dom', {
...
 
+        names = names.split(' ');
+
         var _self = this.__self,
-            selector = '.' +
-                names.split(' ').map(function(name) {
-                    return buildClass(_self._name, name, modName, modVal);
-                }).join(',.');
+            keys = names.map(function(name) {
+                return _self.buildClass(name, modName, modVal);
```
Comment (Yeti-or) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: in bem-bl buildClass is not static: https://github.com/bem/bem-bl/blob/i-bem-dom-caching%40support/2.x/blocks-common/i-bem/__dom/i-bem__dom.js#L55

File: blocks-common/i-bem/__dom/i-bem__dom.js [hunk 3]
```diff
@@ -824,15 +836,8 @@ var DOM = BEM.DOM = BEM.decl('i-bem__dom', {
...
-        var key = name + buildModPostfix(modName, modVal),
-            res;
-
-        if(!(res = this._elemCache[key])) {
-            res = this._elemCache[key] = this.findElem(name, modName, modVal);
-            res.__bemElemName = name;
-        }
-
-        return res;
+        return this._elemCache[this.__self.buildClass(name, modName, modVal)] ||
```
Comment (Yeti-or) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: and here

File: blocks-common/i-bem/__dom/i-bem__dom.js [hunk 4]
```diff
@@ -806,12 +806,24 @@ var DOM = BEM.DOM = BEM.decl('i-bem__dom', {
...
-        return findDomElem(ctx, selector);
+            keys = names.map(function(name) {
+                return _self.buildClass(name, modName, modVal);
+            }),
+            isSingleName = keys.length === 1,
+            selector = '.' + (isSingleName? keys[0] : keys.join(',.')),
+            res = findDomElem(ctx, selector);
+
+        // caching results if possible
+        (ctx.length === 1 && this.domElem.length === 1 && ctx[0] === this.domElem[0]) &&
```
Comment (Yeti-or) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: case: 
element is at other dom node and block has been destructed.
so `ctx.length === 1` - true
`this.domElem.length` - Cannot read property 'length' of undefined
change to `ctx.length === 1 && this.domElem && this.domElem.length === 1`

**Comment by Yeti-or** [comment]
Update docs please. `findElem()` is caching too: https://github.com/bem/bem-bl/blob/i-bem-dom-caching%40support/2.x/blocks-common/i-bem/i-bem.ru.md#finding
And I think guys here will be grateful too:https://ru.bem.info/technology/i-bem/v2/i-bem-js/#dom

**Comment by Yeti-or** [comment]
And you forgot about `dropElemCache`

``` javascript
     dropElemCache: function(names, modName, modVal) {
         if(names) {
-            var modPostfix = buildModPostfix(modName, modVal);
-
-            (names.indexOf(' ') === -1)
-                ? delete this._elemCache[names + modPostfix]
-                : names.split(' ').forEach(function(name) {
-                    delete this._elemCache[name + modPostfix];
-                }, this);
+            names.split(' ').forEach(function(name) {
+                delete this._elemCache[buildClass(this.__self._name, name, modName, modVal)];
+            }, this);
         } else {
             this._elemCache = {};
         }
```

File: blocks-common/i-bem/__dom/i-bem__dom.js (hunk 4 above)
Comment (Yeti-or) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: by the way what do u think if  `res` is empty `res === []` ?

**Comment by Yeti-or** [comment]
and so there is no need in `buildModPostfix = INTERNAL.buildModPostfix`

File: blocks-common/i-bem/__dom/i-bem__dom.js (hunk 4 above)
Comment (qfox) [inline blocks-common/i-bem/__dom/i-bem__dom.js]: It will cache empty result. It seems viable.

Fixed all the rest.

**Comment by qfox** [comment]
> Update docs please. findElem() is caching too: https://github.com/bem/bem-bl/blob/i-bem-dom-caching%40support/2.x/blocks-common/i-bem/i-bem.ru.md#finding

`findElem` storing cache but not caching ;-\ Are you sure?

**Comment by qfox** [comment]
@Yeti-or fixed.

**Comment by Yeti-or** [comment]
@zxqfox thank you for contribution!

**Comment by qfox** [comment]
@Yeti-or hope it helps ;-) U r welcome!

**Pull request merged**